File: Lab 5 (27 February 2025)/Djkstra algorithm.cpp
#include <iostream>
#include <climits>
using namespace std;

#define V 5
int minDistance(int dist[], bool visited[])
{
    int min = INT_MAX, min_index = -1;
    for (int v = 0; v < V; v++)
    {
        if (!visited[v] && dist[v] < min)
        {
            min = dist[v];
            min_index = v;
        }
    }
    return min_index;
}

void dijkstra(int graph[V][V], int src)
{
    int dist[V];
    bool visited[V];
    for (int i = 0; i < V; i++)
    {
        dist[i] = INT_MAX;
        visited[i] = false;
    }
    dist[src] = 0;

    for (int count = 0; count < V - 1; count++)
    {
        int u = minDistance(dist, visited);
        visited[u] = true;
        for (int v = 0; v < V; v++)
        {
            if (!visited[v] && graph[u][v] && dist[u] != INT_MAX &&
                dist[u] + graph[u][v] < dist[v])
                {
                    dist[v] = dist[u] + graph[u][v];
                }
        }
    }
    cout << "Vertex   Distance from Source " << src << "\n";
    for (int i = 0; i < V; i++)
    {
        cout << i << " \t\t " << dist[i] << "\n";
    }
}

int main()
{
    int graph[V][V] = {
        {0, 10, 0, 5, 0},
        {10, 0, 1, 2, 0},
        {0, 1, 0, 0, 4},
        {5, 2, 0, 0, 3},
        {0, 0, 4, 3, 0}
    };
    int src;
    cout << "Enter the source vertex (0 to " << V - 1 << "): ";
    cin >> src;
    dijkstra(graph, src);
    return 0;
}

File: Lab 5 (27 February 2025)/Djkstra algorithm.py
import random
import heapq
import time
import matplotlib.pyplot as p
import math

def dijkstra(graph, source, V):
    dist = [float('inf')] * V
    dist[source] = 0
    pq = [(0, source)]  # (distance, vertex)

    while pq:
        current_dist, u = heapq.heappop(pq)

        if current_dist > dist[u]:
            continue

        for v, weight in graph[u]:
            if dist[v] > current_dist + weight:
                dist[v] = current_dist + weight
                heapq.heappush(pq, (dist[v], v))
    return dist

def generate_graph(V, avg_degree=2):
    graph = [[] for _ in range(V)]
    for u in range(V):
        for _ in range(avg_degree):
            v = random.randint(0, V - 1)
            while v == u:
                v = random.randint(0, V - 1)
            weight = random.randint(1, 10)
            graph[u].append((v, weight))
    return graph

sizes = [100, 200, 500, 1000, 2000, 3000]
times = []

for V in sizes:
    graph = generate_graph(V)
    start_time = time.time()
    dijkstra(graph, 0, V)
    end_time = time.time()
    times.append(end_time - start_time)

k = times[0] / ((sizes[0] * math.log2(sizes[0])))
onlogv = [k * V * math.log2(V) for V in sizes]

p.figure(figsize=(10, 5))
p.plot(sizes, times, marker="o", color="red", label="Measured Time (Dijkstra)")
p.plot(sizes, onlogv, linestyle="--", color="blue", label="O((V+E) log V) Reference")

p.xlabel("Number of Vertices (V)")
p.ylabel("Time (seconds)")
p.title("Time Complexity of Dijkstra's Algorithm")
p.legend()
p.grid(True)
p.show()